File: 21558 - Nithin/week 10/5-longest-palindromic-substring/5-longest-palindromic-substring.cpp
class Solution {
public:
    //Dynamic programming bottom up approach (tabulation)
    string longestPalindrome(string s) {
         int n = s.size();
        int dp[n][n];
        memset(dp,0,sizeof(dp));
        int beg = 0, end = 1;
        
        for(int i = 0;i < n;i++){   //all single characters are by default palindromic
            dp[i][i] = 1;
        }
        
        for(int i = 0;i < n-1;i++){
           if(s[i] == s[i+1]){
               dp[i][i+1] = 1;
               beg = i;
               end = 2;
           } 
        }
        
        for(int j = 2;j < n;j++){
            for(int i = 0;i < n - j;i++){
                int left = i;
                int right = i+j;
                
                if(dp[left+1][right-1] == 1 && s[left] == s[right]){
                    dp[left][right] = 1;
                    beg = i;
                    end = j+1;
                }
            }
        }
        return s.substr(beg, end);
    }
        
};

File: 21551-naveen-siva-kumar/week10/5_Longest Palindromic Substring.py
class Solution {
public:
    string longestPalindrome(string s) {
        int n =s.length();
        int start=0,end=1;
        
        bool dp[n][n];
        memset(dp,0,sizeof(dp));
        int i,j;
        

        for(i=0;i<n;i++){
            dp[i][i]=true;
        }
        for(i=0;i<n-1;i++){
            if(s[i]==s[i+1]){
                dp[i][i+1]=true;
                start = i;
                end=2;
            }
        }

        int right,left;
        for(i=2;i<n;i++){
            for(j=0;j<n-i;j++){
                 left=j;right=i+j;
                if(dp[left+1][right-1]==true && s[left]==s[right]){
                    dp[left][right]=true; 
                    start=j;
                    end=i+1;
                    
                } 
            }
        }

        
        return s.substr(start,end);
    }
};
